src/passes/inlineAssemblyTransformer.ts

```diff
@@ -0,0 +1,106 @@
+import { ASTMapper } from '../ast/mapper';
+import { AST } from '../ast/ast';
+import { createExpressionStatement, createIdentifier, createTuple } from '../utils/nodeTemplates';
+
+import {
+  InlineAssembly,
+  YulNode,
+  Identifier,
+  ASTNode,
+  BinaryOperation,
+  Assignment,
+  Block,
+  Expression,
+  ExpressionStatement,
+  VariableDeclaration,
+  UncheckedBlock,
+} from 'solc-typed-ast';
+import assert from 'assert';
+
+class YulTransformer {
+  vars: { [name: string]: VariableDeclaration };
+  ast: AST;
+
+  constructor(ast: AST, externalReferences: any[], contextMap: Map<number, ASTNode>) {
+    this.ast = ast;
+    const blockRefIds = externalReferences.map((ref) => ref.declaration);
+    const blockRefs: [number, ASTNode][] = [...contextMap].filter(([id, _]) =>
+      blockRefIds.includes(id),
+    );
+    this.vars = Object.fromEntries(
+      blockRefs.map(([id, ref]) => [(ref as Identifier).name, ref as VariableDeclaration]),
+    );
+  }
+
+  run(node: YulNode): ASTNode {
```

Comment:
My initial idea was to be able to easily unit test the class thus the ASTNode type so you could pass any YulNode and receive any ASTNode. However many of util functions require `ast` and `node` so it will be hard to isolate the class anyway. I guess I will just use a Statement as a return type

Reply:
It doesn't feel right to me to use the result of `run` as an Expression or Statement depending on the context. They have different semantic meanings, the result of `run` is either an expression or a statement. Treating this way as it can be what you want depending on where the function was call, in my opinion makes the code unreadable for others and for the future you which won't remember